<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Изучение-данных" data-toc-modified-id="Изучение-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Изучение данных</a></span></li><li><span><a href="#Вывод-по-изучению-данных" data-toc-modified-id="Вывод-по-изучению-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Вывод по изучению данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Предобработка данных</a></span></li><li><span><a href="#Вывод-по-этапу-предобработки-данных:" data-toc-modified-id="Вывод-по-этапу-предобработки-данных:-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Вывод по этапу предобработки данных:</a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Решающее-дерево" data-toc-modified-id="Решающее-дерево-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Решающее дерево</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Выводы</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Взвешивание-классов" data-toc-modified-id="Взвешивание-классов-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Взвешивание классов</a></span></li><li><span><a href="#Upsampling-и-downsampling" data-toc-modified-id="Upsampling-и-downsampling-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Upsampling и downsampling</a></span><ul class="toc-item"><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Выводы</a></span></li></ul></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span><ul class="toc-item"><li><span><a href="#Выводы" data-toc-modified-id="Выводы-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Выводы</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Обучение с учителем
# Проект: Отток клиентов банка
# Описание проекта

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Нужно построить несколько моделей и выбрать одну с предельно большим значением `F1-меры`. Чтобы сдать проект успешно, нужно довести метрику до 0.59 и проверить `F1-меру` лучшей модели на тестовой выборке.

Дополнительно рекомендовано измерять *AUC-ROC* и сравнивать её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score, roc_auc_score
from sklearn.utils import shuffle
from sklearn.preprocessing import RobustScaler

## Подготовка данных
### Изучение данных

Прочтем файл `Churn.csv` и сохраните его в переменной `df`:

In [2]:
try:    
    df = pd.read_csv('D:/notebook/project_7/Churn.csv')   
except:
    df = pd.read_csv('/datasets/Churn.csv')
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

Описание данных

Признаки

* `RowNumber` — индекс строки в данных
* `CustomerId` — уникальный идентификатор клиента
* `Surname` — фамилия
* `CreditScore` — кредитный рейтинг
* `Geography` — страна проживания
* `Gender` — пол
* `Age` — возраст
* `Tenure` — сколько лет человек является клиентом банка
* `Balance` — баланс на счёте
* `NumOfProducts` — количество продуктов банка, используемых клиентом
* `HasCrCard` — наличие кредитной карты (1 - есть, 0 - нет)
* `IsActiveMember` — активность клиента (1 - активный, 0 - нет)
* `EstimatedSalary` — предполагаемая зарплата

Целевой признак

* `Exited` — факт ухода клиента

Получим общую информацию о таблице:

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [4]:
df['Tenure'].isnull().sum()

909

В датафрейме содержатся данные разных типов: float64, int64 и object. В столбце Tenure присутствуют 909 пропущенных значений.

Проверим наличие грубых дубликатов в датафрейме.

In [5]:
df.duplicated().sum()

0

Посмотрим на уникальные данные всех столбцов.

In [6]:
def check_columns(data_frame):
    for column in data_frame.columns:
        print('Уникальные значения столбца', column)
        print(data_frame[column].unique())
        
check_columns(df)

Уникальные значения столбца RowNumber
[    1     2     3 ...  9998  9999 10000]
Уникальные значения столбца CustomerId
[15634602 15647311 15619304 ... 15584532 15682355 15628319]
Уникальные значения столбца Surname
['Hargrave' 'Hill' 'Onio' ... 'Kashiwagi' 'Aldridge' 'Burbidge']
Уникальные значения столбца CreditScore
[619 608 502 699 850 645 822 376 501 684 528 497 476 549 635 616 653 587
 726 732 636 510 669 846 577 756 571 574 411 591 533 553 520 722 475 490
 804 582 472 465 556 834 660 776 829 637 550 698 585 788 655 601 656 725
 511 614 742 687 555 603 751 581 735 661 675 738 813 657 604 519 664 678
 757 416 665 777 543 506 493 652 750 729 646 647 808 524 769 730 515 773
 814 710 413 623 670 622 785 605 479 685 538 562 721 628 668 828 674 625
 432 770 758 795 686 789 589 461 584 579 663 682 793 691 485 650 754 535
 716 539 706 586 631 717 800 683 704 615 667 484 480 578 512 606 597 778
 514 525 715 580 807 521 759 516 711 618 643 671 689 620 676 572 695 592
 567 694 547 594 673 61

Изучим данные, содержащиеся в столбце целевого признака `Exited`:

In [7]:
df.Exited.value_counts(normalize=True)

0    0.7963
1    0.2037
Name: Exited, dtype: float64

Имеется дисбаланс классов 4:1
* Пользуются услугами банка - 0.7963
* Не пользуются услугами банка - 0.2037

Остальные признаки делятся на две группы: 

* Категориальные: `Gender, Geography, HasCrCard и IsActiveMember`
* Количественные: `CreditScore, Age, Tenure, NumOfProducts, Balance и EstimatedSalary`

Посмотрим распределение значений в категориальных признаках.

In [8]:
categorical_list = ['Gender', 'Geography', 'HasCrCard', 'IsActiveMember']
def check_columns(df):
    for column in categorical_list:
        print('Количество значений в столбце', column)
        print(df[column].value_counts())
        
check_columns(df)

Количество значений в столбце Gender
Male      5457
Female    4543
Name: Gender, dtype: int64
Количество значений в столбце Geography
France     5014
Germany    2509
Spain      2477
Name: Geography, dtype: int64
Количество значений в столбце HasCrCard
1    7055
0    2945
Name: HasCrCard, dtype: int64
Количество значений в столбце IsActiveMember
1    5151
0    4849
Name: IsActiveMember, dtype: int64


Посмотрим статистические характеристики количественных признаков.

In [9]:
df[['CreditScore', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'EstimatedSalary']].describe()

CreditScore           Age       Tenure        Balance  NumOfProducts  \
count  10000.000000  10000.000000  9091.000000   10000.000000   10000.000000   
mean     650.528800     38.921800     4.997690   76485.889288       1.530200   
std       96.653299     10.487806     2.894723   62397.405202       0.581654   
min      350.000000     18.000000     0.000000       0.000000       1.000000   
25%      584.000000     32.000000     2.000000       0.000000       1.000000   
50%      652.000000     37.000000     5.000000   97198.540000       1.000000   
75%      718.000000     44.000000     7.000000  127644.240000       2.000000   
max      850.000000     92.000000    10.000000  250898.090000       4.000000   

       EstimatedSalary  
count     10000.000000  
mean     100090.239881  
std       57510.492818  
min          11.580000  
25%       51002.110000  
50%      100193.915000  
75%      149388.247500  
max      199992.480000

### Вывод по изучению данных

1. В датафрейме 10000 строк 14 столбцов, где каждая строка это - клиент, каждый столбец - признак. 3 типа данных: float, int и object. Грубые дубликаты отсутствуют.

2. В столбце `Tenure` есть пропущенные данные (около 10% пропусков). Данные могли отсутствовать изначально или быть утеряны из-за сбоев в системе при выгрузке базы. 

3. Изучены уникальные значения во всех столбцах.

4. Целевым признаком является столбец `Exited`. В выборке представлены клиенты, где 79.6% - остались с банком, а 20.4% ушли от банка, что говорит нам о том, имеется дисбаланс в данных.

5. Признаками содержащими лишнюю информацию для машинного обучения являются: `Surname, RowNumber и CustomerId`.

6. Остальные признаки делятся на две группы: количественные и категориальные.
- Категориальные: `Gender, Geography, HasCrCard и IsActiveMember`
- Количественные: `CreditScore, Age, Tenure, NumOfProducts, Balance и EstimatedSalary`

7. Изучено распределение значений в категориальных  признаках.

8. Изучены статистические характеристики количественных признаков.

### Предобработка данных

Признаки `RowNumber, CustomerId и Surname` не несут в себе информации, необходимой для дальнейшей классификации, удалим их.

In [10]:
df = df.drop(['RowNumber', 'CustomerId', 'Surname'], axis=1)

Заменим пропущенные значения в столбце tenure на ноль, сами данные столбца для обучения нам не понадобятся, а вот терять почти 10% строк не будем.

In [11]:
df['Tenure'] = df['Tenure'].fillna(0)

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   CreditScore      10000 non-null  int64  
 1   Geography        10000 non-null  object 
 2   Gender           10000 non-null  object 
 3   Age              10000 non-null  int64  
 4   Tenure           10000 non-null  float64
 5   Balance          10000 non-null  float64
 6   NumOfProducts    10000 non-null  int64  
 7   HasCrCard        10000 non-null  int64  
 8   IsActiveMember   10000 non-null  int64  
 9   EstimatedSalary  10000 non-null  float64
 10  Exited           10000 non-null  int64  
dtypes: float64(3), int64(6), object(2)
memory usage: 859.5+ KB


Изменим тип данных на `object` в категориальных признаках, в столбцах 'HasCrCard', 'IsActiveMember'.

In [13]:
for column in ['HasCrCard', 'IsActiveMember']:
    df[column] = df[column].astype('object')

Преобразуем категориальные признаки в численные техникой прямого кодирования One-Hot Encoding, используя функцию `pd.get_dummies()`, с аргументом `drop_first=True`, который удалит первоначальный столбец из датафрейма. Это необходимо для удаления связанной информации, которая может привести нас в "дамми-ловушку".

In [14]:
df = pd.get_dummies(df, drop_first=True)

Преобразуем типы данных в столбцах `'CreditScore', 'Age', 'Tenure', 'NumOfProducts', 'EstimatedSalary'` в `float` для экономии места.

In [15]:
for column in ['CreditScore', 'Age', 'Tenure', 'NumOfProducts', 'EstimatedSalary']:
     df[column] = pd.to_numeric(df[column], downcast='float')

Преобразуем типы данных в столбце `'Exited'` в `int` для экономии места.

In [16]:
df['Exited'] = pd.to_numeric(df['Exited'], downcast='integer')

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CreditScore        10000 non-null  float32
 1   Age                10000 non-null  float32
 2   Tenure             10000 non-null  float32
 3   Balance            10000 non-null  float64
 4   NumOfProducts      10000 non-null  float32
 5   EstimatedSalary    10000 non-null  float32
 6   Exited             10000 non-null  int8   
 7   Geography_Germany  10000 non-null  uint8  
 8   Geography_Spain    10000 non-null  uint8  
 9   Gender_Male        10000 non-null  uint8  
 10  HasCrCard_1        10000 non-null  uint8  
 11  IsActiveMember_1   10000 non-null  uint8  
dtypes: float32(5), float64(1), int8(1), uint8(5)
memory usage: 332.2 KB


Взглянем на преображенный датафрейм.

In [18]:
df.head(10)

CreditScore   Age  Tenure    Balance  NumOfProducts  EstimatedSalary  \
0        619.0  42.0     2.0       0.00            1.0    101348.882812   
1        608.0  41.0     1.0   83807.86            1.0    112542.578125   
2        502.0  42.0     8.0  159660.80            3.0    113931.570312   
3        699.0  39.0     1.0       0.00            2.0     93826.632812   
4        850.0  43.0     2.0  125510.82            1.0     79084.101562   
5        645.0  44.0     8.0  113755.78            2.0    149756.703125   
6        822.0  50.0     7.0       0.00            2.0     10062.799805   
7        376.0  29.0     4.0  115046.74            4.0    119346.882812   
8        501.0  44.0     4.0  142051.07            2.0     74940.500000   
9        684.0  27.0     2.0  134603.88            1.0     71725.726562   

   Exited  Geography_Germany  Geography_Spain  Gender_Male  HasCrCard_1  \
0       1                  0                0            0            1   
1       0                  0                1            0            0   
2       1                  0                0            0            1   
3       0                  0                0            0            0   
4       0                  0                1            0            1   
5       1                  0                1            1            1   
6       0                  0                0            1            1   
7       1                  1                0            0            1   
8       0                  0                0            1            0   
9       0                  0                0            1            1   

   IsActiveMember_1  
0                 1  
1                 1  
2                 0  
3                 0  
4                 1  
5                 0  
6                 1  
7                 0  
8                 1  
9                 1

### Вывод по этапу предобработки данных:
На этапе предобработки данных мы выполнили первичную подготовку данных для дальнейшего анализа и машинного обучения:

* Удалили признаки `RowNumber, CustomerId, Surname` из датафрейма.
* Исправили пропущенные значения в столбце `'Tenure'`, заменив их на 0, чтобы не терять порядка 10 % строк датафрейма.
* Изменили тип данных в столбцах для экономии места. 
* Создали  новые столбцы для категориальных признаков, преобразовав их в численные, воспользовавшись техникой прямого кодирования `One-Hot Encoding`.

## Исследование задачи

Для начала разделим данные на признаки и целевой признак

In [19]:
features = df.drop('Exited', axis=1)
target = df['Exited']

Разделим данные на обучающую, валидационную и тестовую выборки в пропорции 3:1:1 соответственно (60% на обучающую и по 20% данных на валидационную и тестовую выборки).

In [20]:
# разделение проведём с помощью функции train_test_split библиотеки sklearn
X, X_test, y, y_test = train_test_split(features, target, test_size=0.2, random_state=12345)
X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.25, random_state=12345)

In [21]:
numeric = ['CreditScore', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'EstimatedSalary']

rs = RobustScaler()


rs.fit(X[numeric])
rs.fit(X_train[numeric])

X[numeric] = rs.transform(X[numeric])
X_test[numeric] = rs.transform(X_test[numeric])

X_train[numeric] = rs.transform(X_train[numeric])
X_valid[numeric] = rs.transform(X_valid[numeric])


In [22]:
X.head()

CreditScore       Age  Tenure   Balance  NumOfProducts  EstimatedSalary  \
454      0.783582  0.166667     0.4  0.236977            0.0        -0.441284   
8307    -0.261194 -0.583333    -0.2  0.287704            0.0        -0.698118   
4811    -0.582090 -0.916667     0.6 -0.759006            1.0         0.129429   
1132    -0.186567  0.000000    -0.8 -0.759006            1.0         0.733506   
9640     0.723881 -0.166667    -0.2 -0.759006            0.0         0.935673   

      Geography_Germany  Geography_Spain  Gender_Male  HasCrCard_1  \
454                   0                0            0            0   
8307                  0                0            1            1   
4811                  0                1            0            1   
1132                  0                0            1            1   
9640                  0                0            0            1   

      IsActiveMember_1  
454                  1  
8307                 1  
4811                 0  
1132                 1  
9640                 0

In [23]:
# выведем размеры получившихся выборок
print('Размеры наборов данных:')
print('X:       ', X.shape)
print('X_train: ', X_train.shape)
print('X_valid: ', X_valid.shape)
print('X_test:  ', X_test.shape)
print('y:       ', y.shape)
print('y_train: ', y_train.shape)
print('y_valid: ', y_valid.shape)
print('y_test:  ', y_test.shape)

Размеры наборов данных:
X:        (8000, 11)
X_train:  (6000, 11)
X_valid:  (2000, 11)
X_test:   (2000, 11)
y:        (8000,)
y_train:  (6000,)
y_valid:  (2000,)
y_test:   (2000,)


Попробуем обучить различные классификационные модели без учёта дисбаланса классов.

### Решающее дерево

* Инициируем модель решающего дерева DecisionTreeClassifier
* Выделим гиперпараметры для настройки модели:
    * max_depth - максимальная глубина древа
    * min_samples_split - минимальное количество объектов в узле
    * min_samples_leaf - минимальное количество объектов в листе
    * criterion ("gini" или "entropy") - функция измерения качества разделения
* Критерием проверки качества модели будет f1 и roc_auc.

In [24]:
def decision_tree(X_train, X_valid, y_train, y_valid, cw=None):
    """
    input: X_train, X_valid, y_train, y_valid - обучающие и валидационные выборки признаков и целевого признака
            cw - параметр class_weight для модели решающего дерева
    output: None
    description: функция обучает модель решающего дерева DecisionTreeClassifier с различными сочетаниями гиперпараметров;
                 метрики качества модели: f1 и roc_auc;
                 функция выводит наилучшие значения метрик и значения гиперпараметров наилучшей модели.
    """
    # в параметрах best_params_f1 и best_f1 будем хранить наилучшие параметры для модели и наивысшую f1 метрику соответственно
    best_params_f1 = []
    best_f1 = 0
    # аналогично для roc_auc метрики.
    best_params_roc_auc = []
    best_roc_auc = 0.5
    # проходим в циклах по всем выбранным гиперпараметрам
    for max_depth in range(2,14):
        for min_samples_split in range(2,8):
            for min_samples_leaf in range(1,10):
                for criterion in ['gini', 'entropy']:
                    # инициация модели с текущими гиперпараметрами
                    model_dt_clf = DecisionTreeClassifier(random_state=12345,
                                                          max_depth=max_depth,
                                                          min_samples_split=min_samples_split,
                                                          min_samples_leaf=min_samples_leaf,
                                                          criterion=criterion,
                                                          class_weight=cw)
                    # обучение модели на тренировочной выборке
                    model_dt_clf.fit(X_train, y_train)
                    # поиск предсказаний модели на валидационной выборке
                    predictions = model_dt_clf.predict(X_valid)
                    # вычисление f1 модели методом f1_score
                    f1 = f1_score(y_valid, predictions)
                    # если текущее значение метрики f1 выше предыдущего лучшего значения, 
                    # сохраняем параметры модели и текущую метрику f1
                    if f1 > best_f1:
                        best_params_f1 = [max_depth, min_samples_split, min_samples_leaf, criterion]
                        best_f1 = f1
                    # аналогично для метрики roc_auc
                    probabilities_valid = model_dt_clf.predict_proba(X_valid)
                    probabilities_one_valid = probabilities_valid[:, 1]
                    roc_auc = roc_auc_score(y_valid, probabilities_one_valid)
                    
                    if roc_auc > best_roc_auc:
                        best_params_roc_auc = [max_depth, min_samples_split, min_samples_leaf, criterion]
                        best_roc_auc = roc_auc                                          
                       
    
    # вывод наилучшей метрики f1 и параметров модели.
    print('''Наилучшая метрика f1: {}
    Параметры модели:
       max_depth = {}
       min_samples_split = {}
       min_samples_leaf = {}
       criterion = {} \n'''.format(round(best_f1, 3),
                                 best_params_f1[0],
                                 best_params_f1[1],
                                 best_params_f1[2],
                                 best_params_f1[3]
                                ))
    
    # вывод наилучшей метрики roc_auc и параметров модели.
    print('''Наилучшая метрика roc_auc: {}
    Параметры модели:
       max_depth = {}
       min_samples_split = {}
       min_samples_leaf = {}
       criterion = {} '''.format(round(best_roc_auc, 3),
                                 best_params_roc_auc[0],
                                 best_params_roc_auc[1],
                                 best_params_roc_auc[2],
                                 best_params_roc_auc[3]
                                ))

In [25]:
%%time
decision_tree(X_train, X_valid, y_train, y_valid)

Наилучшая метрика f1: 0.567
    Параметры модели:
       max_depth = 7
       min_samples_split = 2
       min_samples_leaf = 8
       criterion = gini 

Наилучшая метрика roc_auc: 0.835
    Параметры модели:
       max_depth = 6
       min_samples_split = 2
       min_samples_leaf = 5
       criterion = entropy 
Wall time: 25.5 s


* Наилучшая метрика f1 для решающего дерева 0.567 (недостаточно по техническому заданию: f1>0.59).
* Наилучший результат по метрикам f1 и roc_auc достигается при разной настройке гиперпараметров.
* Попробуем улучшить результат на других моделях.

### Случайный лес

* Инициируем модель случайного леса RandomForestClassifier
* Выделим гиперпараметры для настройки модели:
    * n_estimators - количество деревьев
    * max_depth - максимальная глубина древа
    * min_samples_split - минимальное количество объектов в узле
    * min_samples_leaf - минимальное количество объектов в листе
* Критерием проверки качества модели будет f1 и roc_auc.

In [26]:
def random_forest(X_train, X_valid, y_train, y_valid, cw=None):
    """
    input: X_train, X_valid, y_train, y_valid - обучающие и валидационные выборки признаков и целевого признака
            cw - параметр class_weight для модели случайного леса
    output: None
    description: функция обучает модель решающего дерева RandomForestClassifier с различными сочетаниями гиперпараметров;
                 метрики качества модели: f1 и roc_auc;
                 функция выводит наилучшие значения метрик и значения гиперпараметров наилучшей модели.
    """
    # в параметрах best_params_f1 и best_f1 будем хранить наилучшие параметры для модели и наивысшую f1 метрику соответственно
    best_params_f1 = []
    best_f1 = 0
    # аналогично для roc_auc метрики.
    best_params_roc_auc = []
    best_roc_auc = 0.5
    # проходим в циклах по всем выбранным гиперпараметрам
    for n_estimators in range(30,40):
        for max_depth in range(10,17,2):
            for min_samples_split in range(2,5):
                for min_samples_leaf in range(1,4):
                    # инициация модели с текущими гиперпараметрами
                    model_rf_clf = RandomForestClassifier(random_state=12345,
                                                          n_estimators=n_estimators,
                                                          max_depth=max_depth,
                                                          min_samples_split=min_samples_split,
                                                          min_samples_leaf=min_samples_leaf)
                    # обучение модели на тренировочной выборке
                    model_rf_clf.fit(X_train, y_train)
                    # поиск предсказаний модели на валидационной выборке
                    predictions = model_rf_clf.predict(X_valid)

                    # вычисление f1 модели методом f1_score
                    f1 = f1_score(y_valid, predictions)
                    # если текущее значение метрики f1 выше предыдущего лучшего значения, 
                    # сохраняем параметры модели и текущую метрику f1
                    if f1 > best_f1:
                        best_params_f1 = [n_estimators, max_depth, min_samples_split, min_samples_leaf]
                        best_f1 = f1
                    # аналогично для метрики roc_auc
                    probabilities_valid = model_rf_clf.predict_proba(X_valid)
                    probabilities_one_valid = probabilities_valid[:, 1]
                    roc_auc = roc_auc_score(y_valid, probabilities_one_valid)
                    
                    if roc_auc > best_roc_auc:
                        best_params_roc_auc = [n_estimators, max_depth, min_samples_split, min_samples_leaf]
                        best_roc_auc = roc_auc

    # вывод наилучшей метрики f1 и параметров модели.
    print('''Наилучшая метрика f1: {}
    Параметры модели:
       n_estimators = {}
       max_depth = {}
       min_samples_split = {}
       min_samples_leaf = {} \n'''.format(round(best_f1, 3),
                                 best_params_f1[0],
                                 best_params_f1[1],
                                 best_params_f1[2],
                                 best_params_f1[3]
                                ))

    # вывод наилучшей метрики roc_auc и параметров модели.
    print('''Наилучшая метрика roc_auc: {}
    Параметры модели:
       n_estimators = {}
       max_depth = {}
       min_samples_split = {}
       min_samples_leaf = {} '''.format(round(best_roc_auc, 3),
                                 best_params_roc_auc[0],
                                 best_params_roc_auc[1],
                                 best_params_roc_auc[2],
                                 best_params_roc_auc[3]
                                ))

In [27]:
%%time
random_forest(X_train, X_valid, y_train, y_valid)

Наилучшая метрика f1: 0.585
    Параметры модели:
       n_estimators = 31
       max_depth = 16
       min_samples_split = 2
       min_samples_leaf = 1 

Наилучшая метрика roc_auc: 0.853
    Параметры модели:
       n_estimators = 39
       max_depth = 10
       min_samples_split = 2
       min_samples_leaf = 1 
Wall time: 1min 10s


Наилучшая метрика f1 для случайного леса 0.586 (не удовлетворяет техническое задание: f1>0.59).
Наилучший результат по метрикам f1 и roc_auc достигается при разной настройке гиперпараметров.
Попробуем улучшить результат на других моделях.

### Логистическая регрессия
* Инициируем модель логистической регрессии LogisticRegression
* Критерием проверки качества модели будет f1 и roc_auc.

In [28]:
%%time

model_lr_clf =  LogisticRegression(random_state=12345, solver='liblinear')
model_lr_clf.fit(X_train, y_train)
predictions = model_lr_clf.predict(X_valid)
f1 = f1_score(y_valid, predictions)

probabilities_valid = model_lr_clf.predict_proba(X_valid)
probabilities_one_valid = probabilities_valid[:, 1]
roc_auc = roc_auc_score(y_valid, probabilities_one_valid)

print('Наилучшая метрика f1:', round(f1, 3))
print ('Наилучшая метрика roc_auc:', round(roc_auc,3))

Наилучшая метрика f1: 0.306
Наилучшая метрика roc_auc: 0.771
Wall time: 14.5 ms


* Наилучшая метрика f1 для логистической регрессии 0.306 (хуже, чем на остальных моделях).
* Наилучшая метрика roc_auc 0.771 (хуже, чем на остальных моделях).

Попробуем улучшить результат, разобравшись с дисбалансом классов.

### Выводы
* Исходные данные разделены на обучающую, валидационную и тестовую выборки в соотношении 3:1:1 соответственно
* Обучены модели "Решающее Дерево", "Случайный лес", "Логистическая Регрессия" без учёта дисбаланса классов. Максимальное значение метрики f1 = 0.586 получено при обучении модели случайного леса. Качество модели не удовлетворяет условию технического задания.
* Качество моделей также оценено по метрике roc_auc. Максимальное значение метрики roc_auc = 0.853 достигнуто при обучении модели случайного леса.
* Метрики f1 и roc_auc достигли максимальных значений при одинаковых настройках гиперпараметров.

## Борьба с дисбалансом

* Соотношение значений целевого признака 4:1 (для классов 0 и 1 соответственно)
* Для борьбы с дисбалансом классов попробуем:
    * взвешивание классов (class_weight='balanced')
    * увеличение выборки положительного класса (копирование объектов)
    * уменьшение выборки отрицательного класса (sampling)
    * перемешивание объектов методом shuffle

### Взвешивание классов

In [29]:
%%time
decision_tree(X_train, X_valid, y_train, y_valid, cw='balanced')

Наилучшая метрика f1: 0.571
    Параметры модели:
       max_depth = 7
       min_samples_split = 2
       min_samples_leaf = 3
       criterion = entropy 

Наилучшая метрика roc_auc: 0.844
    Параметры модели:
       max_depth = 6
       min_samples_split = 2
       min_samples_leaf = 9
       criterion = entropy 
Wall time: 26.5 s


In [30]:
%%time
random_forest(X_train, X_valid, y_train, y_valid, cw='balanced')

Наилучшая метрика f1: 0.585
    Параметры модели:
       n_estimators = 31
       max_depth = 16
       min_samples_split = 2
       min_samples_leaf = 1 

Наилучшая метрика roc_auc: 0.853
    Параметры модели:
       n_estimators = 39
       max_depth = 10
       min_samples_split = 2
       min_samples_leaf = 1 
Wall time: 1min 11s


* Взвешивание классов не дало прироста в метрике f1.
* Применим upsampling и downsampling. 

### Upsampling и downsampling

Напишем функции для увеличения и уменьшения выборок.

In [31]:
def upsample(features, target, repeat):
    """
    input:        features - признаки 
                  target - целевой признак
                  repeat (int) - множитель для повторения признаков
    output:       features_upsampled - увеличенная выборка признаков
                  target_upsampled - увеличенная выборка целевого признака
    description:  функция разделяет признаки по классам целевого признака (0, 1),
                  повторяет признаки целевого положительного класса repeat раз,
                  объединяет объекты с разными целевыми признаками методом concat,
                  перемешивает объекты методом shuffle
    """
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]
    
    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)
    return features_upsampled, target_upsampled

In [32]:
def downsample(features, target, fraction):
    """
    input:        features - признаки 
                  target - целево признак
                  fraction (float) - доля объектов для семплирования
    output:       features_downsampled - уменьшенная выборка признаков
                  target_downsampled - уменьшенная выборка целевого признака
    description:  функция разделяет признаки по классам целевого признака (0, 1), 
                  сэмплирует признаки целевого класса 0 (с коэффициентом fraction),
                  объединяет объекты с разными целевыми признаками методом concat,
                  перемешивает объекты методом shuffle
    """
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]
    
    features_downsampled = pd.concat(
        [features_zeros.sample(frac=fraction, random_state=12345)] + [features_ones])
    target_downsampled = pd.concat(
        [target_zeros.sample(frac=fraction, random_state=12345)] + [target_ones])
    
    features_downsampled, target_downsampled = shuffle(
        features_downsampled, target_downsampled, random_state=12345)
    return features_downsampled, target_downsampled

Для обучающей выборки увеличим количество объектов класса 1 в 2 раза, уменьшим количество объектов класса 0 в 1.5 раза.

In [33]:
X_train_upsampled, y_train_upsampled = upsample(X_train, y_train, 2)
X_train_balanced, y_train_balanced = downsample(X_train_upsampled, y_train_upsampled, 0.75)

Обучим модели случайного леса и решающего дерева на сбалансированных обучающих выборках.

In [34]:
%%time
random_forest(X_train_balanced, X_valid, y_train_balanced, y_valid)

Наилучшая метрика f1: 0.617
    Параметры модели:
       n_estimators = 30
       max_depth = 12
       min_samples_split = 2
       min_samples_leaf = 3 

Наилучшая метрика roc_auc: 0.854
    Параметры модели:
       n_estimators = 39
       max_depth = 10
       min_samples_split = 2
       min_samples_leaf = 3 
Wall time: 1min 10s


In [35]:
%%time
decision_tree(X_train_balanced, X_valid, y_train_balanced, y_valid)

Наилучшая метрика f1: 0.593
    Параметры модели:
       max_depth = 8
       min_samples_split = 2
       min_samples_leaf = 7
       criterion = entropy 

Наилучшая метрика roc_auc: 0.839
    Параметры модели:
       max_depth = 6
       min_samples_split = 2
       min_samples_leaf = 8
       criterion = entropy 
Wall time: 24.5 s


* Метрика f1 для модели "Случайный лес" достигла значения 0.616 на сбалансированной выборке, что удовлетворяет техническое задание (>0.59).
* Для модели "Решающее дерево" метрика f1 достигла значения 0.593 на сбалансированной выборке, что удовлетворяет техническое задание (>0.59), за меньшее время обучения.
* Наилучшие значения метрик f1 и roc_auc достигнуты моделью случайного леса при разных значениях гиперпараметров. Для решающего дерева оптимальные значения гиперпараметров также одинаковые.

#### Выводы
* Взвешивание классов с помощью параметра class_weight не дало увеличения метрики f1.
* Upsampling положительного класса и downsampling отрицательного класса дали прирост точности моделей.
* Необходимое качество на валидационной выборке получили модели:
    * Случайный лес:
        * f1 = 0.616
        * roc_auc = 0.854
        * repeat = 2 - повторение объектов класса 1
        * fraction = 0.75 - коэффициент семплирования объектов класса 0
    * Решающее дерево:
        * f1 = 0.593
        * roc_auc = 0.839
        * repeat = 2
        * fraction = 0.75

## Тестирование модели

* Проведём обучение полученных моделей на всех доступных данных (обучающие + валидационные), предварительно сбалансировав классы с найденными ранее коэффициентами repeat и fraction.
* Вычислим метрики f1 и roc_auc на тестовой выборке.

Модель "Случайный лес"

In [36]:
%%time
X_upsampled, y_upsampled = upsample(X, y, 2)
X_balanced, y_balanced = downsample(X_upsampled, y_upsampled, 0.75)
model_rfc_final = RandomForestClassifier(random_state=12345,
                                         n_estimators=30,
                                         max_depth=12,
                                         min_samples_split=2,
                                         min_samples_leaf=3)
model_rfc_final.fit(X_balanced, y_balanced)
predictions = model_rfc_final.predict(X_test)
f1 = f1_score(y_test, predictions)

probabilities_valid = model_rfc_final.predict_proba(X_test)
probabilities_one_valid = probabilities_valid[:, 1]
roc_auc = roc_auc_score(y_test, probabilities_one_valid)

print(' f1 = {}, \n roc_auc = {}'.format(round(f1,3), round(roc_auc,2)))

 f1 = 0.65, 
 roc_auc = 0.86
Wall time: 218 ms


### Выводы
Значение метрики f1 на модели случайный лес:
   * f1 = 0.65 
   * roc_auc = 0.86 

## Общий вывод

* Произведена предобработка данных.
    * В столбце `Tenure` обнаружены пропущенные данные (около 10% пропусков).
    * Целевым признаком является столбец `Exited`. В выборке представлены клиенты, где 79.6% - остались с банком, а 20.4% ушли от банка, что говорит о том, что данные довольно несбалансированные 4:1.
    * Признаки делятся на две группы: количественные и категориальные.
        * Категориальные: `Gender, Geography, HasCrCard и IsActiveMember`
        * Количественные: `CreditScore, Age, Tenure, NumOfProducts, Balance и EstimatedSalary`
    * Удалены признаки `RowNumber, CustomerId, Surname` из датафрейма.
    * Исправили пропущенные значения в столбце `'Tenure'`, заменены на 0, чтобы не терять порядка 10 % строк датафрейма.
    * Изменен тип данных в столбцах для экономии места.
    * Категориальные признаки преобразованы методом `One-Hot Encoder`.
* Проведён анализ моделей без учёта дисбаланса классов.
    * Исходные данные разделены на обучающую, валидационную и тестовую выборки в соотношении `3:1:1` соответственно.
    * Обучены модели "Решающее Дерево", "Случайный лес", "Логистическая Регрессия" без учёта дисбаланса классов. Максимальное значение метрики `f1 = 0.586` получено при обучении модели случайного леса.
    * Качество моделей также оценено по метрике `roc_auc`. Максимальное значение метрики `roc_auc = 0.853` достигнуто при обучении модели случайного леса.
    * Метрики `f1` и roc_auc достигли максимальных значений при разных настройках гиперпараметров.
* Проведён анализ дисбаланса классов.
    * Взвешивание классов с помощью параметра `class_weight` не дало увеличения метрики `f1`.
    * `Upsampling` положительного класса и `downsampling` отрицательного класса дали прирост точности моделей.
    * Значение `f1` на валидационной выборке:
        * Случайный лес:
            * `f1 = 0.616`
            * `roc_auc = 0.854`
        * Решающее дерево:
            * `f1 = 0.593`
            * `roc_auc = 0.839`
* Выполнена проверка `F1-мера` и `AUC-ROC` на тестовой выборке для модели случайны лес:    
  * `f1 = 0.65`
  * `roc_auc = 0.86`      

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: задача исследована
    - [x]  Исследован баланс классов
    - [x]  Изучены модели без учёта дисбаланса
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 3: учтён дисбаланс
    - [x]  Применено несколько способов борьбы с дисбалансом
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 4: проведено тестирование
- [x]  Удалось достичь *F1*-меры не менее 0.59
- [x]  Исследована метрика *AUC-ROC*